docs/architecture/adr-007.md

```diff
@@ -10,88 +13,101 @@ The current way to define those events is not flexible: events are hardcoded, pa
+```solidity
+// Current
+event __CronosSendToIbc(address sender, string recipient, uint256 amount);
 
+// Proposal
+event __CronosSendToIbcV2(address sender, string recipient, uint256 amount, string channelId, bytes extraData);
+```
 
+#### Description
 
-An example of smart contract implementing those function
+- Upgrade the event version for the argument changes
+  - Current event `__CronosSendToIbc` is kept to avoid introducing confusion and avoid extra chain logic to check the emitted event format at runtime.
+- Channel ID argument is added to support sending to specified IBC channel
+  - For token originated from Cronos, this channel ID could be any available IBC channel;
+  - For token originated from IBC channels, the initial implementation can add a restriction to enforce the same channel as the source channel before it is ready.
 
-```
-pragma solidity ^0.6.8;
+### __CronosSendToChain Event
 
-import "ds-token/token.sol";
+#### Changes
 
-Contract Demo is DSToken {
+- Change `chain_id` to `string` type to accept arbitrary chain ID;
+- Add new argument `extraData bytes` to accept extra arguments in the future.
 
- __CosmosCallAfterPayment(string functionName, address sender, uint amount, byte[]args)
- 
- // unsafe_burn burn tokens without user's approval and authentication, used internally
- function unsafe_burn(address addr, uint amount) internal {
-     // Deduct user's balance without approval
-     require(balanceOf[addr] >= amount, "ds-token-insufficient-balance");
-     balanceOf[addr] = sub(balanceOf[addr], amount);
-     totalSupply = sub(totalSupply, amount);
-     emit Burn(addr, amount);
- }
- 
- // send an "amount" of the contract token to recipient through IBC
- // args should contains the encoded value recipient or extra value if needed in the future
- function send_to_ibc(uint amount, byte[] args) public {
-     unsafe_burn(msg.sender, amount);
-     emit __CosmosCallAfterPayment("CronosSendToIbc", msg.sender, amount, args);
- }
+#### Signature
 
+```solidity
+// Current
+event __CronosSendToChain(address sender, address recipient, uint256 amount, uint256 bridge_fee, uint256 chain_id);
+
+// Proposal
+event __CronosSendToChain(address sender, address recipient, uint256 amount, uint256 bridge_fee, string chain_id, extraData bytes);
 ```
 
-In this example byte[] args should contains the encoded recipient address from `event __CronosSendToIbc(address sender, string recipient, uint256 amount)` but in the future, if we needs to add extra parameters such as `fee_amount` or `fee_recipient` we will be able to use the same method in the smart contract, but encode those extra parameters in args. 
+#### Description
 
-On the chain side, the events parameters are decoded and uses to call a specific actions. 
+- Chain ID argument is chagned from unsigned integer to string. This allow sending token to destination chains other than Ethereum (Gravity Bridge).
```

Comment:
for non ethereum chain, should be redefine new types of event instead?

Reply:
For some reason my last comment was gone, let me re-type it

When I first look at `__CronosSendToChain` I thought it was intended for sending to any arbitrary chains. I could think of a few directions for this event:

1. A generic bridge event, pretty much the same for now that includes sender, recipient, amount (with denom), destination + extra data for chain/protocol specific arguments
2. By bridge protocol, `__CronosSendToGravityBridge` just like what `__CronosSendToIBC` is doing. Caveat is token still need to migrate when we add new protocols
3. By destination chain, `__CronosSendToEthereum`, similar to 2.

Reply:
maybe we could rename to `__CronosSendToEVMChain`

For non evm chain, it is not clear what would be the flow or the parameters at the moment, not sure if we can design it now, unless we encapsulate all the parameters in []args except the one used by the contract

Reply:
> maybe we could rename to __CronosSendToEVMChain
And for EVM chain the underlying protocol is via Gravity Bridge right?

Reply:
yes gravity bridge is compatible with any EVM compatible chain.

Reply:
Updated